onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -1923,6 +1923,179 @@ Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-
+          const auto* input_type1 = ctx.getInputType(1);
+          if (nullptr == input_type0 || nullptr == input_type1) {
+            fail_type_inference("At least on of the inputs missing a type");
+          }
+          if (getTensorElementType(*input_type0) != getTensorElementType(*input_type1)) {
+            fail_type_inference("Both inputs must have the same element type");
+          }
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+          if (!hasInputShape(ctx, 0) || !hasInputShape(ctx, 1)) {
+            return;
+          }
+          // Verify input shapes are the same
+          const auto& shape0 = getInputShape(ctx, 0);
+          const auto& shape1 = getInputShape(ctx, 1);
+          if (shape0.dim_size() != shape1.dim_size()) {
+            fail_shape_inference("Both inputs must have the same shape");
+          }
+
+          auto& dims0 = shape0.dim();
+          auto& dims1 = shape1.dim();
+          for (int i = 0, dim_size = shape0.dim_size(); i < dim_size; ++i) {
+            auto& dim0 = dims0.Get(i);
+            auto& dim1 = dims1.Get(i);
+            if (dim0 != dim1) {
+              fail_shape_inference("Input dims at: ", i, " do not match");
+            }
+          }
+          propagateShapeFromInputToOutput(ctx, 0, 0);
+        });
+
+    static const char* Binarizer_doc = R"DOC(
+The operator will accept a single sparse tensor input and will output
+a sparse tensor of the same shape and type. For each of the elements it
+would compare the value of the 'threshold' attribute and for all values above the threshold
+it would set the output value for the same index to T(1) and T(0) otherwise.
+
+XXX: Can output be simply an integer or a boolean?
+)DOC";
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(Binarizer)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "A", "A sparse matrix of numeric type", "T")
+      .Attr(
+        "threshold",
+        "A boundary value",
+        AttributeProto::INT,
+        static_cast<int64_t>(0))
+       .Output(0, "Y", "Matrix multiply results", "T")
+       .TypeConstraint("T", {"sparse_tensor(float)"}, "Constrain input and output float at this time")
+       .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          propagateShapeFromInputToOutput(ctx, 0, 0);
+        });
+
+    static const char* Bucketizer_doc = R"DOC(
+This operator behaves similar to https://pytorch.org/docs/stable/generated/torch.bucketize.html
+XXX: the output would be a sparse tensor of the same shape, but each value in it would be replaced
+to an index of the corresponding bucket.
+XXX: The output type attribute will be omitted for now. Fixed to int64_t.
+)DOC";
+  ONNX_CONTRIB_OPERATOR_SCHEMA(Bucketizer)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "A", "A sparse matrix of numeric type", "T")
+      .Input(0, "B", "1-D Boundaries tensor", "T1")
```

Comment:
About the boundary format:
- is it possible to support overlapped buckets?
- can we support half-open buckets at left and right end, i.e. the leftmost bucket is between ( -Infinity,  boundary[0] ] and the rightmost [ boundary[last index], +Infinity ) ?